pkg/oauth/apiserver/apiserver.go

```diff
@@ -95,49 +144,103 @@ func (c *completedConfig) newV1RESTStorage() (map[string]rest.Storage, error) {
-		return nil, fmt.Errorf("error building REST storage: %v", err)
+		return nil, nil, fmt.Errorf("error building REST storage: %v", err)
 	}
 	userOAuthAccessTokensDelegate, err := useroauthaccesstokensdelegate.NewREST(accessTokenStorage)
 	if err != nil {
-		return nil, fmt.Errorf("error building REST storage: %v", err)
+		return nil, nil, fmt.Errorf("error building REST storage: %v", err)
 	}
+	tokenReviewStorage, tokenReviewPostStartHooks, err := c.tokenReviewStorage(corev1Client, oauthClient, userClient)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error building REST storage: %v", err)
+	}
+
+	v1Storage := map[string]rest.Storage{
+		"oAuthAuthorizeTokens":      authorizeTokenStorage,
+		"oAuthAccessTokens":         accessTokenStorage,
+		"oAuthClients":              clientStorage,
+		"oAuthClientAuthorizations": clientAuthorizationStorage,
+		"userOAuthAccessTokens":     userOAuthAccessTokensDelegate,
+		"tokenReviews":              tokenReviewStorage,
+	}
+	return v1Storage, tokenReviewPostStartHooks, nil
+}
+
+func (c *completedConfig) tokenReviewStorage(
+	corev1Client corev1.CoreV1Interface,
+	oauthClient *oauthclients.Clientset,
+	userClient *userclient.Clientset,
+) (rest.Storage, map[string]genericapiserver.PostStartHookFunc, error) {
+	bootstrapUserDataGetter := bootstrap.NewBootstrapUserDataGetter(corev1Client, corev1Client)
+
+	// add indexes to the userinformer for the users to be used in user <-> groups mapping
+	userInformer := c.ExtraConfig.UserInformers
+	if err := userInformer.User().V1().Groups().Informer().AddIndexers(cache.Indexers{
+		usercache.ByUserIndexName: usercache.ByUserIndexKeys,
+	}); err != nil {
+		return nil, nil, err
+	}
+
+	groupMapper := usercache.NewGroupCache(userInformer.User().V1().Groups())
+	oauthInformer := c.ExtraConfig.OAuthInformers
+
+	timeoutValidator := tokenvalidators.NewTimeoutValidator(
+		oauthClient.OauthV1().OAuthAccessTokens(),
+		oauthInformer.Oauth().V1().OAuthClients().Lister(),
+		c.ExtraConfig.AccessTokenInactivityTimeout,
+		minimumInactivityTimeoutSeconds)
+
+	postStartHooks := map[string]genericapiserver.PostStartHookFunc{}
+	postStartHooks["openshift.io-StartUserInformer"] = func(ctx genericapiserver.PostStartHookContext) error {
+		go userInformer.Start(ctx.StopCh)
+		return nil
+	}
+	postStartHooks["openshift.io-StartOAuthInformer"] = func(ctx genericapiserver.PostStartHookContext) error {
+		go oauthInformer.Start(ctx.StopCh)
+		return nil
+	}
+	postStartHooks["openshift.io-StartTokenTimeoutUpdater"] = func(ctx genericapiserver.PostStartHookContext) error {
+		go timeoutValidator.Run(ctx.StopCh)
+		return nil
+	}
+
+	// create the authenticator for the "normal" OpenShift OAuth2 users
+	tokenAuthenticator := tokenvalidation.NewTokenAuthenticator(
+		oauthClient.OauthV1().OAuthAccessTokens(), userClient.UserV1().Users(), groupMapper, c.ExtraConfig.APIAudiences,
+		tokenvalidators.NewExpirationValidator(), tokenvalidators.NewUIDValidator(), timeoutValidator)
```

Comment:
the previous `validators := []oauth.OAuthTokenValidator{oauth.NewExpirationValidator(), oauth.NewUIDValidator(), timeoutValidator}` was easier to read